ccl/kvccl/kvtenantccl: TestTenantUpgradeFailure failed

ccl/kvccl/kvtenantccl.TestTenantUpgradeFailure [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3323130&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3323130&tab=artifacts#/) on master @ [df6603d0093fbb7f801f287b3126f11dabc6b9d5](https://github.com/cockroachdb/cockroach/commits/df6603d0093fbb7f801f287b3126f11dabc6b9d5):


```
I210819 02:14:51.290667 8843 1@server/server.go:993  [n?] 492  monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I210819 02:14:51.298722 8843 server/init.go:196  [n?] 493  no stores initialized
I210819 02:14:51.299458 8843 server/init.go:197  [n?] 494  awaiting ``cockroach init`` or join with an already initialized node
I210819 02:14:51.299778 8843 server/init.go:261  [n?] 495  cluster caf50646-87ad-4bc4-9664-3deb123600c7 has been created
I210819 02:14:51.300054 8843 server/init.go:262  [n?] 496  allocated node ID: n1 (for self)
I210819 02:14:51.300327 8843 server/init.go:263  [n?] 497  active cluster version: 21.1
I210819 02:14:51.301288 8843 gossip/gossip.go:402  [n1] 498  NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:33241" > attrs:<> locality:<tiers:<key:"region" value:"test" > tiers:<key:"dc" value:"dc1" > > ServerVersion:<major_val:21 minor_val:1 patch:0 internal:140 > build_tag:"v21.2.0-alpha.00000000-3425-gdf6603d009" started_at:1629339291301282172 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.0.0.1:33473" >
I210819 02:14:51.305007 9166 1@server/server.go:1560  [n1] 499  connecting to gossip network to verify cluster ID "caf50646-87ad-4bc4-9664-3deb123600c7"
I210819 02:14:51.305274 9170 kv/kvserver/closedts/provider/provider.go:135  [ct-closer] 500  disabling legacy closed-timestamp mechanism; the new one is enabled
I210819 02:14:51.305425 8843 server/node.go:403  [n1] 501  initialized store s1
I210819 02:14:51.305469 8843 kv/kvserver/stores.go:250  [n1] 502  read 0 node addresses from persistent storage
I210819 02:14:51.305532 8843 server/node.go:480  [n1] 503  started with engine type 2
I210819 02:14:51.305556 8843 server/node.go:482  [n1] 504  started with attributes []
I210819 02:14:51.305635 8843 1@server/server.go:1687  [n1] 505  starting https server at 127.0.0.1:41799 (use: 127.0.0.1:41799)
I210819 02:14:51.305668 8843 1@server/server.go:1692  [n1] 506  starting postgres server at 127.0.0.1:33473 (use: 127.0.0.1:33473)
I210819 02:14:51.305713 8843 1@server/server.go:1694  [n1] 507  starting grpc server at 127.0.0.1:33241
I210819 02:14:51.305739 8843 1@server/server.go:1695  [n1] 508  advertising CockroachDB node at 127.0.0.1:33241
I210819 02:14:51.307239 9166 1@server/server.go:1563  [n1] 509  node connected via gossip
W210819 02:14:51.346119 9140 kv/range_lookup.go:243  [n1,rangefeed=table-stats-cache,range-lookup=/Meta2/Table/20/NULL] 510  range lookup of key /Meta2/Table/20/NULL found only non-matching ranges []; retrying
W210819 02:14:51.349758 9139 kv/range_lookup.go:243  [n1,rangefeed=table-stats-cache,range-lookup=/Table/20] 511  range lookup of key /Table/20 found only non-matching ranges []; retrying
I210819 02:14:51.353044 9147 kv/kvserver/replica_rangefeed.go:635  [n1,rangefeed=table-stats-cache,s1,r16/1:/Table/2{0-1}] 512  RangeFeed closed timestamp is empty
I210819 02:14:51.361240 8843 1@util/log/event_log.go:32  [n1] 513 ={"Timestamp":1629339291361236479,"EventType":"node_join","NodeID":1,"StartedAt":1629339291301282172,"LastUp":1629339291301282172}
I210819 02:14:51.361318 8843 sql/sqlliveness/slinstance/slinstance.go:276  [n1] 514  starting SQL liveness instance
I210819 02:14:51.366876 9326 sql/sqlliveness/slstorage/slstorage.go:393  [n1] 515  inserted sqlliveness session 07ffcb82ea8341aca2a4c4536d5820c2
I210819 02:14:51.366956 9326 sql/sqlliveness/slinstance/slinstance.go:168  [n1] 516  created new SQL liveness session 07ffcb82ea8341aca2a4c4536d5820c2
I210819 02:14:51.368991 9328 sql/temporary_schema.go:491  [n1] 517  running temporary object cleanup background job
I210819 02:14:51.371045 9328 sql/temporary_schema.go:528  [n1] 518  found 0 temporary schemas
I210819 02:14:51.371147 9328 sql/temporary_schema.go:531  [n1] 519  early exiting temporary schema cleaner as no temporary schemas were found
I210819 02:14:51.371188 9328 sql/temporary_schema.go:532  [n1] 520  completed temporary object cleanup job
I210819 02:14:51.371208 9328 sql/temporary_schema.go:611  [n1] 521  temporary object cleaner next scheduled to run at 2021-08-19 02:44:51.363044192 +0000 UTC
I210819 02:14:51.394452 9464 util/log/event_log.go:32  [n1,intExec=optInToDiagnosticsStatReporting] 522 ={"Timestamp":1629339291376539872,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"diagnostics.reporting.enabled\" = true","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-optInToDiagnosticsStatReporting","SettingName":"diagnostics.reporting.enabled","Value":"true"}
I210819 02:14:51.400264 9470 migration/migrationmanager/manager.go:110  [n1,intExec=set-setting,migration-mgr] 523  no need to migrate, cluster already at newest version
I210819 02:14:51.435707 9470 util/log/event_log.go:32  [n1,intExec=set-setting] 524 ={"Timestamp":1629339291398118678,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING version = $1","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-set-setting","PlaceholderValues":["'21.1'"],"SettingName":"version","Value":"21.1"}
I210819 02:14:51.472514 9896 util/log/event_log.go:32  [n1,intExec=initializeClusterSecret] 525 ={"Timestamp":1629339291467343678,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"cluster.secret\" = gen_random_uuid()::STRING","Tag":"SET CLUSTER SETTING","User":"root","ApplicationName":"$ internal-initializeClusterSecret","SettingName":"cluster.secret","Value":"5ba205a6-a31a-4297-9901-183147249122"}
I210819 02:14:51.479912 9942 5@util/log/event_log.go:32  [n1,intExec=create-default-DB] 526 ={"Timestamp":1629339291475987049,"EventType":"create_database","Statement":"CREATE DATABASE IF NOT EXISTS defaultdb","Tag":"CREATE DATABASE","User":"root","DescriptorID":50,"ApplicationName":"$ internal-create-default-DB","DatabaseName":"defaultdb"}
I210819 02:14:51.502568 9948 5@util/log/event_log.go:32  [n1,intExec=create-default-DB] 527 ={"Timestamp":1629339291480305426,"EventType":"create_database","Statement":"CREATE DATABASE IF NOT EXISTS postgres","Tag":"CREATE DATABASE","User":"root","DescriptorID":51,"ApplicationName":"$ internal-create-default-DB","DatabaseName":"postgres"}
I210819 02:14:51.545589 8843 server/server_sql.go:1071  [n1] 528  done ensuring all necessary startup migrations have run
I210819 02:14:51.546102 10192 jobs/job_scheduler.go:349  [n1] 529  waiting 4m0s before scheduled jobs daemon start
I210819 02:14:51.548489 10194 server/auto_upgrade.go:44  [n1] 530  auto upgrade disabled by testing
I210819 02:14:51.589550 9184 3@vendor/github.com/cockroachdb/pebble/event.go:617  [n1,pebble,s1] 531  [JOB 3] WAL created 000004
I210819 02:14:51.594924 8843 1@server/server.go:2322  [n1] 532  serving sql connections
I210819 02:14:51.595764 8843 testutils/testcluster/testcluster.go:1153  [-] 533  WaitForFullReplication
I210819 02:14:51.595814 8843 testutils/testcluster/testcluster.go:1157  [-] 534  WaitForFullReplication took: 494ns
=== CONT  TestTenantUpgradeFailure
    tenant_upgrade_test.go:356: Leaked goroutine: goroutine 13759 [chan receive]:
        github.com/cockroachdb/cockroach/pkg/ccl/kvccl/kvtenantccl_test.TestTenantUpgradeFailure.func3.2(0xc0004dae40)
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go:344 +0x34
        created by github.com/cockroachdb/cockroach/pkg/ccl/kvccl/kvtenantccl_test.TestTenantUpgradeFailure.func3
        	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go:343 +0xbb9
--- FAIL: TestTenantUpgradeFailure (8.26s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantUpgradeFailure PKG=./pkg/ccl/kvccl/kvtenantccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/kv fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgradeFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go
+++ b/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go
@@ -339,14 +339,22 @@ func TestTenantUpgradeFailure(t *testing.T) {
 			defer cleanup()
 			initialTenantRunner = sqlutils.MakeSQLRunner(tca)
 		}
-		// Just resume the upgrade this time.
+		// Keep trying to resume the stopper channel until the channel is closed,
+		// since we may repeatedly wait on it due to transaction retries. In
+		// the other case the stopper is used, so no such risk exists.
 		go func() {
-			<-tenantStopperChannel
+			for {
+				_, ok := <-tenantStopperChannel
+				if !ok {
+					return
+				}
+			}
 		}()
 		// Upgrade the tenant cluster.
 		initialTenantRunner.Exec(t,
 			"SET CLUSTER SETTING version = $1",
 			clusterversion.TestingBinaryVersion.String())
+		close(tenantStopperChannel)
 		// Validate the target version has been reached.
 		initialTenantRunner.CheckQueryResults(t, "SELECT * FROM t", [][]string{{"1"}, {"2"}})
 		initialTenantRunner.CheckQueryResults(t, "SHOW CLUSTER SETTING version",

